File: scripts/test_security.py
#!/usr/bin/env python3
from copy import deepcopy
import importlib.util
from pathlib import Path


module_path = Path(__file__).with_name("validate-roadmap.py")
spec = importlib.util.spec_from_file_location("validate_roadmap", module_path)
validate_roadmap = importlib.util.module_from_spec(spec)
assert spec.loader is not None
spec.loader.exec_module(validate_roadmap)


def valid_data():
    return {
        "pills": [{"label": "Live", "href": "https://example.com/#live", "style": "is-live"}],
        "sections": [
            {
                "id": "live",
                "label": "Live",
                "title": "Running today",
                "icon": "live",
                "tiers": [
                    {
                        "layout": "products",
                        "cards": [{"title": "Card", "body": "[ok](https://example.com)"}],
                    }
                ],
            }
        ],
        "footer": "Footer",
    }


def assert_invalid(mutator, expected):
    data = valid_data()
    mutator(data)
    errors = validate_roadmap.validate(data)
    assert any(expected in error for error in errors), errors


assert validate_roadmap.validate(valid_data()) == []
assert_invalid(lambda data: data["pills"][0].update({"href": "javascript:alert(1)"}), "$.pills[0].href")
assert_invalid(lambda data: data["pills"][0].update({"href": "data:text/html,evil"}), "$.pills[0].href")
assert_invalid(lambda data: data["sections"][0].update({"id": "Not Safe"}), "$.sections[0].id")
assert_invalid(lambda data: data.update({"extra": True}), "$.extra: unknown field")
assert_invalid(lambda data: data["sections"][0]["tiers"][0]["cards"][0].update({"extra": True}), "unknown field")
assert_invalid(lambda data: data["pills"][0].update({"style": "is-evil"}), "$.pills[0].style")
assert_invalid(lambda data: data["sections"][0].pop("icon"), "missing required field 'icon'")
assert_invalid(lambda data: data["sections"][0]["tiers"][0]["cards"][0].update({"body": "[bad](javascript:alert(1))"}), "Markdown link must use http/https URL")

print("security validation tests passed")

File: scripts/test-security.mjs
import assert from "node:assert/strict"
import { buildHtml } from "../lib/render.js"

const baseData = {
  pills: [
    { label: "Safe", href: "https://example.com/path?x=1&y=2", style: "is-live" },
    { label: "Bad", href: "javascript:alert(1)", style: "is-next" },
  ],
  sections: [
    {
      id: "live",
      label: "Live",
      title: "Running today",
      icon: "live",
      tiers: [
        {
          layout: "products",
          cards: [
            {
              title: "Markdown links",
              body: "[ok](https://example.com/a?b=1&c=2) [inject](https://example.com/\"onmouseover=alert(1)) [star](https://example.com/*path*) [js](javascript:alert(1)) [data](data:text/html,evil)",
            },
          ],
        },
      ],
    },
  ],
  footer: "Footer",
}

const html = buildHtml(baseData)

assert.match(html, /href="https:\/\/example\.com\/path\?x=1&amp;y=2"/)
assert.match(html, /href="https:\/\/example\.com\/a\?b=1&amp;c=2"/)
assert.match(html, /href="https:\/\/example\.com\/\*path\*"/)
assert.doesNotMatch(html, /\sonclick="/)
assert.doesNotMatch(html, /\sonmouseover=/)
assert.doesNotMatch(html, /href="[^"]*<em>/)
assert.doesNotMatch(html, /javascript:/)
assert.doesNotMatch(html, /data:text\/html/)
assert.equal((html.match(/target="_blank"/g) || []).length, 2)
assert.match(html, /<a class="rm-pill is-next">Bad<\/a>/)

console.log("security renderer tests passed")
